decisions/0001-bundle-granularity.rst

```diff
@@ -0,0 +1,199 @@
+different versions of the same Bundle, which might be useful for things like
+shared code libraries.
+
+The notifications aspect of Bundles would be handled by a M:M mapping of Bundles
+(not BundleVersions), with a relationship of ``(lending_bundle, borrowing_bundle,
+still_using)``, where ``still_using`` is a boolean flag that is true if the most
+recent version of the borrowing Bundle still has a link to any version of the
+lending bundle.
+
+Changeset-Level Metadata
+========================
+
+Data about who changed what and why are stored at the BundleVersion level of
+granularity. This does mean that a Changelog for a Course would covered many
+different blocks, and that finding when a particular block changed would be
+more difficult. However, it gives a good high level view of the history of the
+work as a whole (the Course) in a simple way. In the Content Library case, each
+problem would be its own Bundle, so we'd still have tracking at the individual
+problem level.
+
+OLX-Aware Metadata
+==================
+
+Data about the OLX contents of a Bundle (e.g. tagging, search indexing) would
+need to be associated to the tuple of (BundleVersion, XBlock Directory). Asset
+association also happens at this level of granularity.
+
+Publishing Transactions
+=======================
+
+When updating a course, a new Snapshot would be created by a Draft, with all the
+changes for any given set of XBlocks. Once the Snapshot successfully completes,
+a new BundleVersion of the Course would be created to point to that Snapshot. An
+interruption in the publishing process during Snapshot creation could result in
+an orphaned Snapshot that's no pointed to by a BundleVersion, but shouldn't
+result in a broken or inconsitent state.
+
+Hierarchy Representation
+========================
+
+The list of XBlock content is represented as a flat set of directories, and all
+navigational hierarchy has to be interpreted by OLX-aware clients. A student's
+path through a course is an LMS/Compositor level concern that structural OLX is
+an input to, and the storage layer of Blockstore shouldn't need to model it.
+
+Reusing Containers
+==================
+
+To re-use a Unit, you would first make a Link to the Bundle where that Unit came
+from, and then specify the Link prefix before the directory where the Unit comes
+from::
+
+    <xblock-include definition="arch_hangout_videos/video+blockstore_overview"/>
+
+This allows for arbitrary reuse at varying levels of granularity.
+
+Intentionality and Tracking Re-use
+==================================
+
+A drawback of approach is that it requires more explicit intentionality in
+course design in order for its BundleVersion dependency tracking to be
+meaningful. You could borrow a single leaf block or container block from another
+Course, but Blockstore itself would only know that the link between the
+BundleVersions existed, not the specific items that were used. This problem is
```

Comment:
> You could borrow a single leaf block or container block from another Course, but Blockstore itself would only know that the link between the BundleVersions existed, not the specific items that were used.

I guess that's not ideal, but what specific problems would it cause?

We could reduce this by allowing Links to specify a subfolder of the upstream bundle, and only that subfolder would appear in the downstream bundle, which provide some useful data and optimizations without going as far as my proposal in making "Folders" a feature-rich entity of their own.

Reply:
I think it's mostly about controlling expectations and how people build on it over time. I'd prefer to draw a harder line and say that the core of Blockstore just says whether you're allowed to reuse things from a Course and tracks updates to said Course (or more granular Bundle), and that anything more specific than that has to be tracked and reported by something with more OLX awareness.

My concern is that if we start building even things like directories, we'll start having to eventually bring OLX parsing and awareness into the core of Blockstore. Adding a subfolder field to this proposal can't answer the question of "what does it mean to link to that Unit", since it doesn't understand how to parse the things inside that Unit to get the leaves.

Even a smart, hierarchical directory tracking system like HOBC introduces complexities and is still only an approximation of re-use tracking. CCX might re-use a directory but only an extremely limited subset of the contents. A randomized problem from a problem bank (v2, not the Content Library of today) might specify an extremely limited range of content from a very large bank. You could link in an HTML module but only because you want that one licensed picture of Anant with a chainsaw. You might want to answer the question not of how many times my thing is reused, but how many courses outside my organization use it.

I think that OLX-aware content re-use tracking is a separate system, and it's going to need to build its own rich data model to answer the questions it cares about. But if we build a barebones solution in Blockstore that's tightly coupled with low level storage and update tracking, then at every step of the way, we'll be pushed into a corner where the far cheaper path is to twist the data model a little more and add just a little feature there, optimize this access just a bit... and then we'll have the same situation we had in the LMS with Grading and continue to have with Scheduling.

Reply:
Yeah, I'm not opposed to what you're suggesting at all, in fact I'm a bit surprised that I don't see more problems with it, which is why I was asking.

What's most important for blockstore is that we always know when something has been updated upstream, but "false positives" are probably fine (e.g. you're linking to an external course to get that picture of Anant; it's critical that you get updates if that picture is modified but it's not a big problem if you get updates when other parts of the external course are also modified, especially if there's a nice course-level changelog and/or the blockstore client that's using this data can act on the notification in a smart way.)

If we forget what I said about allowing links to specify a subfolder, what do you think about blockstore maintaining a hash of not just the contents of each file but also some kind of hash at the folder level? (It could be computed in real-time simply by iterating over the metadata of all files in the bundle and hashing their hashes, with room for optimization later). That way, even though Blockstore doesn't know that you're only using the one image from one unit/folder within that course, a client which does know which folder you're using can trivially compare the folder hash of the old and new version to decide whether or not to prompt you about the update notification it just received from Blockstore.

Reply:
I think that might be a useful optimization to make, but I'd want to test it to see if it's necessary. As you point out, we already have metadata (including hashes) for all files in the Bundle. The two most common use cases I can think of are "show me all the things that changed between that snapshot and this one" and "did this one file I care about change?", which I suspect will work acceptably fast even on large courses.

But since this would be an optimized calculation over immutable data, I think it's something we can add easily add to the format down the line.